packages/test-interface-criteria/src/criteria/Criteria.js

```diff
@@ -13,29 +13,17 @@ module.exports = class Criteria {
 		return opportunity['@type'] || opportunity['type'];
 	}
 
-	getOffers(opportunity) {
-		return opportunity.offers || (opportunity.superEvent && opportunity.superEvent.offers) || []; // Note FacilityUse does not have bookable offers, as it does not allow inheritance
-	}
-
-	getBookableOffers(opportunity) {
-		const offers = this.getOffers(opportunity);
-		return offers ? offers.filter(x =>
-			(!x.availableChannel || x.availableChannel.includes("https://openactive.io/OpenBookingPrepayment"))
-			&& x.advanceBooking != "https://openactive.io/Unavailable"
-			&& (!x.validFromBeforeStartDate || moment(opportunity.startDate).subtract(moment.duration(x.validFromBeforeStartDate)).isBefore())
-		) : [];
-	}
-
 	getRemainingCapacity(opportunity) {
 		return opportunity.remainingAttendeeCapacity !== undefined ? opportunity.remainingAttendeeCapacity : opportunity.remainingUses;
 	}
 
-	testMatch(opportunity) {
-		const matchesCriteria = true;
-		const unmetCriteriaDetails = [];
-		
-		return { matchesCriteria, unmetCriteriaDetails };
-	}
+	get opportunityConstraints() {
+    return {};
+  }
+
+  get offerConstraints() {
+    return {}
+  }
```

Comment:
Interesting! Here's some rationale, but very much welcome your thoughts!

So they're effectively returning constant values that don't need to be calculated each time they're called. Using getters reflects this intention.

Ideally we'd use [public field declarations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Public_field_declarations) for this kind of thing, however they're [not here yet](https://tc39.es/proposal-class-fields/). This looks like the right optimisation when they're available though?

If optimisation was required in the meantime I was thinking either lazy getters, or some other form of memoization, e.g. using proxy handlers, or forcing the public methods of the package to accept the criteria name as a string and handling it there?), but the MVP of just using getters was easiest to start with.

Also worth noting that the same pattern is used in `ReportGenerator`.

Another option considered was to do this manually with objects instead of classes, though that would effectively be manually implementing class inheritance logic, which seems bad?

Style question:
- Do we always prefer functions for consistency, even for simple cases? ([AirBnB have this opinion](https://github.com/airbnb/javascript#accessors--no-getters-setters))

So options:
a) Leave this as getters on the basis that anything that in the future will use public field declarations will be represented as a getter for now (but change the project's style going forward so that we don't use getters and setters for other cases?)
b) Update to functions in the interim, for total consistency?
c) Use an object-based approach rather than classes?

Reply:
@nickevansuk thanks for explaining. I realise that I misinterpreted when I said that the result was different at different times because that logic is bound in a function. The constraint just returns the function, which is a constraint.

I would highly agree with the object-based approach. I'm a believer in only providing as much structure as is needed. With any more, it's easy to complicate things in the future. e.g. right now the point of inheritance is to extend the constraints (` { ...super.offerConstraints }`). However, inheritance also leads the way to inheriting `name`, for example, if an extended class forgets to implement `get name() { }` (which cannot be enforced in JS). Similarly, methods like `getId()` _could_ be overwritten in the future in one of the classes, meaning that it's sub-classes would now have this new functionality. Neither of these outcomes make sense - we want each criteria to have a distinct name and `getId()` is just a static function - it does not depend on which criteria.

Additionally, classes can have multiople instances, which is why you're talking about things like lazy getters or memoization (or singletons, etc). I feel like this adds a lot of entities to what are - to me - just data

Therefore, a simpler structure might be:

`Criteria.d.ts`

```ts
export type Criteria = {
  name: string,
  opportunityConstraints: {
    [ constraintName: string ]: (opportunity: any) => boolean,
  },
  offerConstraints: {
    [ constraintName: string ]: (offer: any, opportunity: any) => boolean,
  },
};
```

`criteriaUtils.js`

These are just functions:

```js
function getId(opportunity) {
  return opportunity['@id'] || opportunity['id'];
}

function getType(opportunity) {
  return opportunity['@type'] || opportunity['type'];
}
```

and then each criteria file would be like:

```js
/**
 * @type {import('./Criteria').Criteria}
 */
const TestOpportunityBookable = {
  name: 'TestOpportunityBookable',
  opportunityConstraints: { ... CriteriaFutureScheduledOpportunity.opportunityConstraints. ... },
  ...
};
```

I think the class structure makes more sense before TypeScript because you want some way of saying "these are all the same kind of thing and they all have these fields". But a TypeScript type is, I think, a much simpler way of doing that

Reply:
Agreed with the object-based approach here